source/math/FloatVectorList.ooc

```diff
@@ -381,4 +382,76 @@ FloatVectorList: class extends VectorList<Float> {
 			result add(this[i] clamp(floor, ceiling))
 		result
 	}
+	findMaxIndex: func -> Int {
+		index := 0
+		for (i in 1 .. this count)
+			if (this[index] < this[i])
+				index = i
+		index
+	}
+	crossCorrelation: func (other: This, vectorLength: Int) -> Float {
+		upperPart := 0.0f
+		lowerPart1 := 0.0f
+		lowerPart2 := 0.0f
+		thisMean := this mean
+		otherMean := other mean
+		thisPointer := (this pointer as Float*)
+		otherPointer := (other pointer as Float*)
+		for (i in 0 .. vectorLength) {
+			thisValue := thisPointer[i]
+			otherValue := otherPointer[i]
+			thisDiff := (thisValue - thisMean)
+			otherDiff := (otherValue - otherMean)
+			upperPart += thisDiff * otherDiff
+			lowerPart1 += thisDiff * thisDiff
+			lowerPart2 += otherDiff * otherDiff
+		}
+		upperPart / sqrt(lowerPart1 * lowerPart2)
+	}
+	shift: func (offset: Int, filledValue := 0.0f) -> This {
+			result := This new(this count)
+			thisPointer := (this pointer as Float*)
+			for (i in 0 .. this count) {
+				index := i + offset
+				value := (index >= 0 && index < this count) ? thisPointer[index] : filledValue
+				result add(value)
+			}
+			result
+	}
+	// offset < 0 means it shifts to left
+	findCrossCorrelationOffset: func (other: This, offsetStart, offsetEnd: Int) -> CorrelationData {
+		crossCorrelation := This new(this count)
+		for (offset in offsetStart .. offsetEnd)
+			crossCorrelation add(this shift(offset) crossCorrelation(other, other count))
+
+		index := crossCorrelation findMaxIndex()
+		maxValue := crossCorrelation[index]
+		index = (index < (offsetEnd - offsetStart) / 2) ? (index + offsetStart) : (index - offsetEnd)
+
+		CorrelationData new(index, maxValue, crossCorrelation)
```

Comment:
We discussed a bit IRL and realized that maybe we don't have to return the vectorlist at all here, since we are actually trying to calculate the offset.